Community info page

darrenvong opened the pull request:
<!---
Please describe the pull request below.
If it closes an issue, make sure to write "closes #1234"
If there is an issue but it isn't completely closed, still refer to the issue number, eg. "part of #1234"
--->
Half of #927. Community edit page to come in the next PR as this is getting bigger than expected (although a good 1/3 of the changes are tests/Storybook code).

**Desktop**
![Screenshot 2021-06-05 at 04 04 17](https://user-images.githubusercontent.com/13669362/120878317-7d7b9100-c5b3-11eb-968d-c247ea5679ad.png)

**Mobile**
![Screenshot 2021-06-05 at 04 03 53](https://user-images.githubusercontent.com/13669362/120878323-840a0880-c5b3-11eb-818b-267322b318dc.png)


<!---
Checklists - you can remove one that is not applicable (ie. remove backend checklist if you only worked on frontend)
If you need help with any of these, please ask :)
--->
**Frontend checklist**
- [x] Formatted my code with `yarn format && yarn lint --fix`
- [x] There are no warnings from `yarn lint`
- [x] There are no console warnings when running the app
- [x] Added any new components to storybook
- [x] Added tests where relevant
- [x] All tests pass
- [x] Clicked around my changes running locally and it works
- [x] Checked Desktop, Mobile and Tablet screen sizes


<!---
Remember to request review from couchers-org/frontend, couchers-org/@backend or an individual.
Once your code is approved, remember to merge it if you have write access
--->

darrenvong committed:
feat: add community info page section and info page skeleton

Refactored the content summary logic into a separate file since it's
shared between discussion card and this

darrenvong committed:
refactor: discussion card to use new logic

Technically unrelated but it's a small change so I thought I'll do it
here...

darrenvong committed:
feat: add community leader and other missing icons to Storybook

darrenvong committed:
feat: show community admins/leaders on info page

darrenvong committed:
refactor: useListAdmins hook to encapsulate more logic

Community leaders into its own component as it's getting more complicated

darrenvong committed:
feat: split community admins query into two

It's basically the same query duplicated to a different query key to
allow the dialog to load more admins without the main one closing it.
Before, it was sharing the same query so fetching more from the dialog
would also affect the UI on the underlying page.

darrenvong committed:
feat: allow all moderators to be seen in a separate dialog

darrenvong committed:
feat: add Storybook entry for community info page

darrenvong committed:
feat: add tests for community info page

lucaslcode commented on app/frontend/src/features/communities/CommunityInfoPage.test.tsx:
```diff
@@ -0,0 +1,192 @@
+      await waitForElementToBeRemoved(screen.getByRole("progressbar"));
+
+      userEvent.click(screen.getByRole("button", { name: SEE_ALL_MODERATORS }));
+      await screen.findByRole("presentation");
```
Does this (and similar) need to be more specific? Match the title or something?

darrenvong replied:
I tried and it didn't work, ends up with a cannot find element with `role="presentation"` :( Might be because the dialog itself also a role "none" on it that interferes with the *ByRole stuff...

darrenvong replied:
Never mind, doing it on role `dialog` works instead, so yes better to match with the title so we know it's the right one 👍

lucaslcode commented on app/frontend/src/features/communities/CommunityInfoPage.test.tsx:
```diff
@@ -0,0 +1,192 @@
+      ).not.toBeInTheDocument();
+    });
+
+    it("closes the dialog by pressing the escape key", async () => {
```
Isn't this testing mui?

darrenvong replied:
I thought about this, but in some places like the bug report dialog we explicitly disable some things like no clicking background for closing, so I guess this is checking this hasn't been disabled by accident

lucaslcode commented on app/frontend/src/features/communities/CommunityInfoPage.tsx:
```diff
@@ -0,0 +1,48 @@
+  },
+  content: {
+    "& p": {
+      margin: 0,
```
Why is this needed here and not the markdown component in general?

darrenvong replied:
I haven't checked if this will break in other places and if we want to disable the margin too, my guess is probably yes? I'll move it into `Markdown` if that's the case 👍

lucaslcode commented on app/frontend/src/features/communities/CommunityLeadersSection.tsx:
```diff
@@ -0,0 +1,92 @@
+          <Alert severity="error">{error.message}</Alert>
+        ) : isLoading ? (
+          <CircularProgress />
+        ) : adminIds && adminIds.length > 0 ? (
```
Time to stop destructuring react-query? Or you prefer this way?

darrenvong replied:
I modified the `useListAdmins` hook slightly so in this case it's returning more than the `listAdmins` query object, and since `adminIds` isn't from the main query object, it doesn't make a difference in this case.

But yeah, definitely worth trying out the type narrowing trick we read on the other post more for these kind of pattern more for a bog standard query object!

lucaslcode commented on app/frontend/src/features/communities/CommunityLeadersSection.tsx:
```diff
@@ -0,0 +1,92 @@
+  const [isModeratorsDialogOpen, setIsModeratorsDialogOpen] = useState(false);
+
+  return (
+    <>
```
Unneeded fragment

lucaslcode commented on app/frontend/src/features/communities/CommunityPage/EventsSection.tsx:
```diff
@@ -15,7 +15,7 @@ import { useCommunityPageStyles } from "./CommunityPage";
 import SectionTitle from "./SectionTitle";
 
-export default function PlacesSection({
+export default function EventsSection({
```
oops

lucaslcode commented on app/frontend/src/features/communities/constants.ts:
```diff
@@ -2,6 +2,7 @@ import { CommunityTab } from "routes";
 
 export const communityTabBarLabels: Record<CommunityTab, string> = {
   overview: "Overview",
+  info: "General",
```
```suggestion
  info: "Local info",
```

lucaslcode commented on app/frontend/src/features/communities/constants.ts:
```diff
@@ -14,6 +15,10 @@ export const CLOSE = "Close";
 export const COMMENT = "Comment";
 export const COMMENTS = "Comments";
 export const COMMUNITY_HEADING = (name: string) => `Welcome to ${name}!`;
+export const COMMUNITY_LEADERS = "Community leaders";
+export const COMMUNITY_LEADERS_DESCRIPTION =
+  "Below is a list of current leaders within this community.";
```
Is this from the desktop mockup? "Leaders" sounds pretty weird to me...

darrenvong replied:
Yeah, and then it has another heading on "Community moderators" right underneath that. Maybe I should move "Community moderators" into the h2 title and remove this description? It's kinda obvious from looking at the UI with what the title says right?

lucaslcode replied:
Yes definitely

darrenvong committed:
feat: fix label and unused fragment from feedback

Also refactored key press into test helper, since usually keydown
accompanies keyup

lucaslcode commented on app/frontend/src/features/communities/CommunityPage/InfoPageSection.tsx:
```diff
@@ -0,0 +1,43 @@
+        <Link
+          to={routeToCommunity(community.communityId, community.slug, "info")}
+        >
+          <Button component="span">{SEE_MORE_INFORMATION}</Button>
```
Do you think this makes more sense as an MuiLink? (And I guess also the other "See more" buttons)

darrenvong replied:
Hmm yeah, since it's changing the user's URL (regardless of with or without React Router) so a link makes more sense to me. Maybe I'll start changing it here...

darrenvong committed:
feat: remove reference to the "community leaders" phrasing

darrenvong committed:
feat: swap "see more X" buttons to links

lucaslcode commented on app/frontend/src/components/Markdown.tsx:
```diff
@@ -27,7 +27,7 @@ const useStyles = makeStyles((theme) => ({
     "& h5": theme.typography.h5,
     "& h6": theme.typography.h6,
-    "& p": theme.typography.body1,
+    "& p": { ...theme.typography.body1, margin: 0 },
```
On second thought, it's better with the paragraph margin. Why did you want to get rid of it in the local info?

darrenvong replied:
I thought it made the gap between the paragraph and the title a little too big, and the fact it's closer to the moderators section makes it looks like it belongs more to that section rather than the "general information" heading immediately below.

**With margin**
![Screenshot 2021-06-06 at 16 18 16](https://user-images.githubusercontent.com/13669362/120929910-d85fd600-c6e2-11eb-949f-a2a52a353e24.png)

**Without margin (current code)**
![Screenshot 2021-06-06 at 16 18 23](https://user-images.githubusercontent.com/13669362/120929912-dc8bf380-c6e2-11eb-9743-a6e35e208297.png)

darrenvong replied:
~~Maybe a better solution is to push the "community moderators" further down?~~ Never mind, pushing the section down doesn't really help after more glancing at it, I think the thing that bothers me is the lack of symmetry between the distance of the paragraph from the page header compared to the Mui parapgraph under "community moderators"

The problem is the page title already has some padding, so that's why having the paragraph also with the margin makes the gap look massive when compared to the normal "body1" paragraph under "community moderators", so actually I think the specific component override is fine? I just checked again and removing the margin makes the markdown paragraph a bit too close to the back button on a discussion page description now...

lucaslcode replied:
I don't mind either way for the headings, but I'm thinking about when you have two paragraphs together, there should be some spacing. If it's set to 0, then it's equivalent to `<br>` and it shouldn't be

lucaslcode replied:
@darrenvong ☝️

lucaslcode reviewed (changes requested):
Looks good, one tiny thing

darrenvong committed:
feat: tweak margins on the pesky markdown paragraph

darrenvong commented:
By the way if you want to get this in soon, feel free to hit merge if you are happy as I might not get around doing it today

lucaslcode reviewed (changes requested):
Failing test

darrenvong committed:
fix: broken tests due to heading change

darrenvong committed:
feat: fix section title always defaulting to h1 for everything

lucaslcode reviewed (changes requested):
Merge conflicts, and did you see my comment about the `<p>` margin?

darrenvong commented:
> did you see my comment about the `<p>` margin?

Yes, I ended up adding `& p` selector so it shares the same `marginTop` that we get in other heading component as well so should fix the issue of two paragraphs being right below each other you mentioned? I didn't need to hack around the margin stuff as much in the end by using `SectionTitle` and making it more flexible to accept different heading levels so it's still semantically correct depending on where it's used.

darrenvong committed:
fix: user summary to use the right props from new changes

Plus the missed proto imports...
